pkg/prometheus/agent/daemonset.go

```diff
@@ -0,0 +1,240 @@
+	// We set some defaults if some fields are not present, and we want those fields set in the original Prometheus object before building the DaemonSetSpec.
+	p.SetCommonPrometheusFields(cpf)
+
+	spec, err := makeDaemonSetSpec(p, config, cg, tlsSecrets)
+	if err != nil {
+		return nil, fmt.Errorf("make DaemonSet spec: %w", err)
+	}
+	daemonSet := &appsv1.DaemonSet{Spec: *spec}
+
+	operator.UpdateObject(
+		daemonSet,
+		operator.WithName(name),
+		operator.WithAnnotations(objMeta.GetAnnotations()),
+		operator.WithAnnotations(config.Annotations),
+		operator.WithLabels(objMeta.GetLabels()),
+		operator.WithLabels(map[string]string{
+			prompkg.PrometheusNameLabelName: objMeta.GetName(),
+			prompkg.PrometheusModeLabeLName: prometheusMode,
+		}),
+		operator.WithLabels(config.Labels),
+		operator.WithManagingOwner(p),
+		operator.WithoutKubectlAnnotations(),
+	)
+
+	if cpf.ImagePullSecrets != nil && len(cpf.ImagePullSecrets) > 0 {
+		daemonSet.Spec.Template.Spec.ImagePullSecrets = cpf.ImagePullSecrets
+	}
+
+	if cpf.HostNetwork {
+		daemonSet.Spec.Template.Spec.DNSPolicy = v1.DNSClusterFirstWithHostNet
+	}
+
+	return daemonSet, nil
+}
+
+func makeDaemonSetSpec(
+	p monitoringv1.PrometheusInterface,
+	c *prompkg.Config,
+	cg *prompkg.ConfigGenerator,
+	tlsSecrets *operator.ShardedSecret,
+) (*appsv1.DaemonSetSpec, error) {
+	cpf := p.GetCommonPrometheusFields()
+
+	pImagePath, err := operator.BuildImagePathForAgent(
+		ptr.Deref(cpf.Image, ""),
+		c.PrometheusDefaultBaseImage,
+		operator.StringValOrDefault(cpf.Version, operator.DefaultPrometheusVersion),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	if !slices.Contains(cpf.EnableFeatures, "agent") {
+		cpf.EnableFeatures = append(cpf.EnableFeatures, "agent")
+	}
+
+	promArgs := buildAgentArgs(cpf, cg)
+
+	volumes, promVolumeMounts, err := prompkg.BuildCommonVolumes(p, tlsSecrets)
+	if err != nil {
+		return nil, err
+	}
+
+	configReloaderVolumeMounts := prompkg.CreateConfigReloaderVolumeMounts()
+
+	var configReloaderWebConfigFile string
+
+	confArg, configVol, configMount, err := prompkg.BuildWebconfig(cpf, p)
+	if err != nil {
+		return nil, err
+	}
+
+	promArgs = append(promArgs, confArg)
+	volumes = append(volumes, configVol...)
+	promVolumeMounts = append(promVolumeMounts, configMount...)
+
+	configReloaderWebConfigFile = confArg.Value
+	configReloaderVolumeMounts = append(configReloaderVolumeMounts, configMount...)
+
+	startupProbe, readinessProbe, livenessProbe := prompkg.MakeProbes(cpf, cg)
+
+	podAnnotations, podLabels := prompkg.BuildPodMetadata(cpf, cg)
+	// In cases where an existing selector label is modified, or a new one is added, new sts cannot match existing pods.
+	// We should try to avoid removing such immutable fields whenever possible since doing
+	// so forces us to enter the 'recreate cycle' and can potentially lead to downtime.
+	// The requirement to make a change here should be carefully evaluated.
```

Comment:
(nit)
```suggestion
	// In cases where an existing selector label is modified, or a new one is added, new daemonset cannot match existing pods.
	// We should try to avoid removing such immutable fields whenever possible since doing
	// so forces us to enter the 'recreate cycle' and can potentially lead to downtime.
	// The requirement to make a change here should be carefully evaluated.
```